doc/src/tutorial/codegen_explainatory.rst

```diff
@@ -0,0 +1,458 @@
+``numpy``. Using the numpy library gives the generated function access to 
+powerful vectorized ufuncs that are backed by compiled C code.
+
+Let us compare the speeds::
+
+    >>> from sympy.utilities.autowrap import ufuncify
+    >>> from sympy.utilities.lambdify import lambdify
+    >>> fn_numpy=lambdify(x, expr,'numpy')
+    >>> fn_fortran=ufuncify([x], expr)
+    >>> from numpy import linspace
+    >>> xx=linspace(0,1,5)
+    >>> fn_numpy(xx)
+    [ 0.          1.21306132  0.98101184  0.44626032  0.        ]
+    >>> fn_fortran(xx)
+    [ 0.          1.21306132  0.98101184  0.44626032  0.        ]
+    >>> import timeit
+    >>> timeit.timeit('fn_numpy(xx)', 'from __main__ import fn_numpy, xx', number=10000)
+    0.176117897034
+    >>> timeit.timeit('fn_numpy(xx)',
+    ...     'from __main__ import fn_numpy, xx', number=10000)
+    0.158110141754
+    >>> timeit.timeit('fn_fortran(xx)',
+    ...     'from __main__ import fn_fortran, xx', number=10000)
+    0.0108680725098
+
+The options available with ufuncify are more or less the same as those 
+available with ``autowrap``.
+
+Codegen (sympy.utilities.codegen)
+---------------------------------
+This module deals with creating compilable code from SymPy expressions. 
+This is lower level than autowrap, as it doesn't actually attempt to 
+compile the code, but higher level than the printers, as it generates 
+compilable files (including header files), rather than just code snippets.
+
+The user friendly functions, here, are ``codegen`` and ``make_routine``.
+``codegen`` takes a list of ``(variable, expression)`` pairs and a language 
+(C, F95, and Octave/Matlab are supported). It returns, as strings, a code 
+file and a header file (for relevant languages). The variables are created 
+as functions that return the value of the expression as output.
+
+.. note:: The ``codegen`` callable is not in the sympy namespace automatically,
+   to use it you must first import ``codegen`` from ``sympy.utilities.codegen``
+
+For instance::
+
+    >>> from sympy.utilities.codegen import codegen
+    >>> from sympy import symbols
+    >>> length, breadth, height=symbols('length, breadth, height')
+    >>> [(c_name, c_code), (h_name, c_header)]=codegen(('volume', length*breadth*height),"C","test",header=False,empty=False)
+    >>> print(c_name)
+    test.c
+    >>> print(c_code)
+    #include "test.h"
+    #include <math.h>
+    double volume(double breadth, double height, double length) {
+       double volume_result;
+       volume_result = breadth*height*length;
+       return volume_result;
+    }
+    >>> print(h_name)
+    test.h
+    >>> print(c_header)
+    #ifndef PROJECT__TEST__H
+    #define PROJECT__TEST__H
+    double volume(double breadth, double height, double length);
+    #endif
+
+Various flags to ``codegen`` let you modify things like the project name
+and global variables.
+
+    name_expr : tuple, or list of tuples
```

Comment:
The following seems to be a copy and paste from the codgen doc string. Is that necessary since you can find it elsewhere? I prefer your style above where you introduced different args and than showed an example.